magellan-library/src/main/java/com/wealthfront/magellan/navigation/NavigationTraverser.kt

```diff
@@ -0,0 +1,34 @@
+package com.wealthfront.magellan.navigation
+
+import com.wealthfront.magellan.core.Journey
+import com.wealthfront.magellan.core.Navigable
+
+class NavigationTraverser(private val root: Journey<*>) {
+
+  fun getGlobalBackStack(): List<Navigable> {
+    val backStackList = mutableListOf<Navigable>()
+    exploreBackStack(root, backStackList)
+    return backStackList.toList()
+  }
+
+  fun getVisibleNavigables(): List<Navigable> {
+    val globalBackStack = getGlobalBackStack()
+    return globalBackStack.filter { it.view != null }
+  }
+
+  private fun exploreBackStack(journey: Journey<*>, backStackList: MutableList<Navigable>) {
+    val backStack = getBackStackForJourney(journey)
+    if (backStack.isEmpty()) return
+    backStack.forEach { navEvent ->
+      val nav = navEvent.navigable
+      backStackList.add(nav)
+      if (nav is Journey<*>) {
+        exploreBackStack(nav, backStackList)
+      }
+    }
+  }
+
+  private fun getBackStackForJourney(journey: Journey<*>): List<NavigationEvent> {
+    return journey.children.mapNotNull { it as? Navigator }.first().backStack
```

Comment:
will add a lint rule to ship with our library to make sure that each journey has just one navigator and also i suppose we need to check that nothing outside of a journey has a navigator as well.

Reply:
Talked offline. We decided that if we wanted to do nested/child navigation, we could always make a child journey. And we _do_ want to enforce the condition of each journey having just one backstack/navigation.